server/request_pusher.go

```diff
@@ -0,0 +1,81 @@
+package server
+
+import (
+	"fmt"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/flashbots/rpc-endpoint/database"
+	"time"
+)
+
+type RequestPusher struct {
+	db         database.Store
+	EntryChan  chan database.Entry
+	BatchChan  chan []database.Entry
+	maxItem    int           // maximum item to be pushed as batch
+	maxTimeOut time.Duration // timeout in which entry to be added to batch queue
+}
+
+func NewRequestPusher(db database.Store, maxItem int, maxTimeOut time.Duration) *RequestPusher {
+	return &RequestPusher{
+		db:         db,
+		EntryChan:  make(chan database.Entry, 100),
+		BatchChan:  make(chan []database.Entry),
+		maxItem:    maxItem,
+		maxTimeOut: maxTimeOut,
+	}
+}
+
+func (r *RequestPusher) Run() {
+	go r.saveEntries()
+	go r.addEntryToEntryQueue()
+}
+
+func (r *RequestPusher) addEntryToEntryQueue() {
+	for keepGoing := true; keepGoing; {
+		var batch []database.Entry
+		ticker := time.NewTicker(r.maxTimeOut)
+		for {
+			select {
+			case entry, ok := <-r.EntryChan:
+				if !ok {
+					keepGoing = false
+					goto push
+				}
+				batch = append(batch, entry)
+				if len(batch) > r.maxItem {
+					goto push
+				}
+			case <-ticker.C:
+				goto push
+			}
+		}
+	push:
+		if len(batch) > 0 {
+			r.BatchChan <- batch
+		}
+	}
+}
+func (r *RequestPusher) saveEntries() {
+	for entry := range r.BatchChan {
```

Comment:
I'm not sure if limiting to a single inserting thread won't result in degraded performance - the bigger the batch the longer you spend waiting for it and there's only so much items you can process at once.
Would just spinning up 4 goroutines doing this work? If so let's keep it in mind.

Reply:
I thought about using workerpool up front. then decided to keep one to see the performance.